File: cabot_people/script/framos_initialize_node.py
#!/usr/bin/env python3

# Copyright (c) 2025  Carnegie Mellon University, IBM Corporation, and others
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

from collections import deque
import signal
import sys

import rclpy
import rclpy.node
import std_srvs.srv
import tf2_ros

from sensor_msgs.msg import CameraInfo


class FramosInitialize():

    def __init__(self, node, map_frame_name, camera_link_frame_name, rgb_camera_topic_name, depth_camera_topic_name, min_rgb_fps, min_depth_fps,
                max_buffer_size, wait_time_after_camera_ready, wait_time_after_tf_ready):
        self.node = node
        self.logger = node.get_logger()

        self.map_frame_name = map_frame_name
        self.camera_link_frame_name = camera_link_frame_name
        self.rgb_camera_topic_name = rgb_camera_topic_name
        self.depth_camera_topic_name = depth_camera_topic_name
        self.min_rgb_fps = min_rgb_fps
        self.min_depth_fps = min_depth_fps
        self.max_buffer_size = max_buffer_size
        self.wait_time_after_camera_ready = wait_time_after_camera_ready
        self.wait_time_after_tf_ready = wait_time_after_tf_ready

        self.is_ready = False
        self.is_reset_running = False
        self.time_tf_ready = None
        self.time_camera_ready = None

        self.rgb_times = deque(maxlen=self.max_buffer_size)
        self.depth_times = deque(maxlen=self.max_buffer_size)

        self.tf2_buffer = tf2_ros.Buffer()
        self.tf2_listener = tf2_ros.TransformListener(self.tf2_buffer, self.node)

        self.rgb_sub = self.node.create_subscription(CameraInfo, self.rgb_camera_topic_name, self.rgb_cb, 10)
        self.depth_sub = self.node.create_subscription(CameraInfo, self.depth_camera_topic_name, self.depth_cb, 10)

        self.enable_client = self.node.create_client(std_srvs.srv.SetBool, 'enable')
        self.enable_client.wait_for_service()

        self.node.create_timer(1.0, self.check_transform)

    def call_enable_client(self, data):
        request = std_srvs.srv.SetBool.Request()
        request.data = data
        return self.enable_client.call_async(request)

    def rgb_cb(self, msg):
        now = rclpy.time.Time.from_msg(msg.header.stamp)
        if self.time_camera_ready is None:
            self.time_camera_ready = now
        elif (now - self.time_camera_ready).nanoseconds * 1.0e-9 > self.wait_time_after_camera_ready:
            self.rgb_times.append(now)

    def depth_cb(self, msg):
        now = rclpy.time.Time.from_msg(msg.header.stamp)
        if self.time_camera_ready is None:
            self.time_camera_ready = now
        elif (now - self.time_camera_ready).nanoseconds * 1.0e-9 > self.wait_time_after_camera_ready:
            self.depth_times.append(now)

    def reset_framos(self):
        def enable_done_callback(future_enable):
            result = future_enable.result()
            if result.success:
                self.logger.info("Successed to enable FRAMOS")
            else:
                self.logger.warn("Failed to enable FRAMOS")
            self.is_reset_running = False

        def disable_done_callback(future_disable):
            result = future_disable.result()
            if result.success:
                self.logger.info("Successed to disable FRAMOS")

                future_enable = self.call_enable_client(True)
                future_enable.add_done_callback(enable_done_callback)
            else:
                self.logger.warn("Failed to disable FRAMOS")
                self.is_reset_running = False

        self.is_reset_running = True

        self.time_camera_ready = None
        self.rgb_times = deque(maxlen=self.max_buffer_size)
        self.depth_times = deque(maxlen=self.max_buffer_size)

        future_disable = self.call_enable_client(False)
        future_disable.add_done_callback(disable_done_callback)

    def check_transform(self):
        if self.is_reset_running:
            return

        try:
            t = self.tf2_buffer.lookup_transform(self.map_frame_name, self.camera_link_frame_name, rclpy.time.Time(seconds=0, nanoseconds=0, clock_type=self.node.get_clock().clock_type))
            if self.time_tf_ready is None:
                self.time_tf_ready = self.node.get_clock().now()
        except (tf2_ros.LookupException, tf2_ros.ConnectivityException, tf2_ros.ExtrapolationException):
            self.logger.info("TF is not ready")
            self.time_tf_ready = None
            return

        now = self.node.get_clock().now()
        if (len(self.rgb_times)==self.rgb_times.maxlen) and (len(self.depth_times)==self.depth_times.maxlen):
            rgb_fps = self.rgb_times.maxlen / ((now - self.rgb_times[0]).nanoseconds * 1.0e-9)
            depth_fps = self.depth_times.maxlen / ((now - self.depth_times[0]).nanoseconds * 1.0e-9)

            if (rgb_fps >= self.min_rgb_fps) and (depth_fps >= self.min_depth_fps):
                self.logger.info("FRAMOS is ready, RGB FPS: %.2f, depth FPS: %.2f" % (rgb_fps, depth_fps))
                self.is_ready = True
            else:
                self.logger.warn("Reset FRAMOS because FPS is low, RGB FPS: %.2f, depth FPS: %.2f" % (rgb_fps, depth_fps))
                self.reset_framos()
        elif (now - self.time_tf_ready).nanoseconds * 1.0e-9 > self.wait_time_after_tf_ready:
            self.logger.warn("Reset FRAMOS because FRAMOS seems not working")
            self.reset_framos()


def main():
    rclpy.init()
    node = rclpy.create_node("framos_initialize")
    map_frame_name = node.declare_parameter('map_frame', 'map').value
    camera_link_frame_name = node.declare_parameter('camera_link_frame', 'camera_link').value
    rgb_camera_topic_name = node.declare_parameter('rgb_camera_topic_name', 'color/camera_info').value
    depth_camera_topic_name = node.declare_parameter('depth_camera_topic_name', 'aligned_depth_to_color/camera_info').value
    min_rgb_fps = node.declare_parameter('min_rgb_fps', 10.0).value
    min_depth_fps = node.declare_parameter('min_depth_fps', 10.0).value
    max_buffer_size = node.declare_parameter('max_buffer_size', 100).value
    wait_time_after_camera_ready = node.declare_parameter('wait_time_after_camera_ready', 10.0).value
    wait_time_after_tf_ready = node.declare_parameter('wait_time_after_tf_ready', 100.0).value

    framos_initialize = FramosInitialize(node, map_frame_name, camera_link_frame_name, rgb_camera_topic_name, depth_camera_topic_name, min_rgb_fps, min_depth_fps,
                                        max_buffer_size, wait_time_after_camera_ready, wait_time_after_tf_ready)

    while rclpy.ok():
        if framos_initialize.is_ready:
            sys.exit(0)
        rclpy.spin_once(node)


def receiveSignal(signal_num, frame):
    print("Received:", signal_num)
    sys.exit(0)


signal.signal(signal.SIGINT, receiveSignal)

if __name__ == "__main__":
    main()

File: cabot_people/src/framos_initialize_node.cpp
// Copyright (c) 2025  IBM Corporation
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include "framos_initialize_node.hpp"


namespace cabot_people
{

FramosInitializeNode::FramosInitializeNode(const rclcpp::NodeOptions & options)
: rclcpp::Node("framos_initialize_node", options),
  map_frame_name_("map"),
  camera_link_frame_name_("camera_link"),
  rgb_camera_topic_name_("color/camera_info"),
  depth_camera_topic_name_("aligned_depth_to_color/camera_info"),
  min_rgb_fps_(10.0),
  min_depth_fps_(10.0),
  max_buffer_size_(100),
  min_wait_after_camera_ready_(5.0),
  max_wait_after_tf_ready_(100.0),
  is_ready_(false),
  time_tf_ready_(0.0),
  time_camera_ready_(0.0)
{
  map_frame_name_ = this->declare_parameter("map_frame", map_frame_name_);
  camera_link_frame_name_ = this->declare_parameter("camera_link_frame", camera_link_frame_name_);
  rgb_camera_topic_name_ = this->declare_parameter("rgb_camera_topic_name", rgb_camera_topic_name_);
  depth_camera_topic_name_ = this->declare_parameter("depth_camera_topic_name", depth_camera_topic_name_);
  min_rgb_fps_ = this->declare_parameter("min_rgb_fps", min_rgb_fps_);
  min_depth_fps_ = this->declare_parameter("min_depth_fps", min_depth_fps_);
  max_buffer_size_ = this->declare_parameter("max_buffer_size", max_buffer_size_);
  min_wait_after_camera_ready_ = this->declare_parameter("min_wait_after_camera_ready", min_wait_after_camera_ready_);
  max_wait_after_tf_ready_ = this->declare_parameter("max_wait_after_tf_ready", max_wait_after_tf_ready_);

  tf_buffer_ = new tf2_ros::Buffer(get_clock());
  tf_listener_ = new tf2_ros::TransformListener(*tf_buffer_, this);

  rgb_sub_ = this->create_subscription<sensor_msgs::msg::CameraInfo>(rgb_camera_topic_name_, 10, std::bind(&FramosInitializeNode::rgb_cb, this, std::placeholders::_1));
  depth_sub_ = this->create_subscription<sensor_msgs::msg::CameraInfo>(depth_camera_topic_name_, 10, std::bind(&FramosInitializeNode::depth_cb, this, std::placeholders::_1));

  set_parameters_client_ = this->create_client<rcl_interfaces::srv::SetParametersAtomically>("set_parameters_atomically");
  set_parameters_client_->wait_for_service();

  timer_ = create_wall_timer(
    std::chrono::duration<double>(1.0),
    std::bind(&FramosInitializeNode::check_transform, this));
}

FramosInitializeNode::~FramosInitializeNode()
{
  delete tf_listener_;
  delete tf_buffer_;
}

bool FramosInitializeNode::is_ready()
{
  return is_ready_;
}

void FramosInitializeNode::rgb_cb(const sensor_msgs::msg::CameraInfo::SharedPtr msg)
{
  double now = rclcpp::Time(msg->header.stamp).seconds();
  if (time_camera_ready_ <= 0) {
    time_camera_ready_ = now;
  } else if ((now - time_camera_ready_) > min_wait_after_camera_ready_) {
    std::lock_guard<std::mutex> rgb_times_lock(rgb_times_mutex_);
    if (rgb_times_.size() >= max_buffer_size_) {
      rgb_times_.pop();
    }
    rgb_times_.push(now);
  }
}

void FramosInitializeNode::depth_cb(const sensor_msgs::msg::CameraInfo::SharedPtr msg)
{
  double now = rclcpp::Time(msg->header.stamp).seconds();
  if (time_camera_ready_ <= 0) {
    time_camera_ready_ = now;
  } else if ((now - time_camera_ready_) > min_wait_after_camera_ready_) {
    std::lock_guard<std::mutex> depth_times_lock(depth_times_mutex_);
    if (depth_times_.size() >= max_buffer_size_) {
      depth_times_.pop();
    }
    depth_times_.push(now);
  }
}

void FramosInitializeNode::reset_framos()
{
  auto enable_done_callback = [this](rclcpp::Client<rcl_interfaces::srv::SetParametersAtomically>::SharedFuture response) {
    if (response.valid() && response.get()->result.successful) {
      RCLCPP_INFO(this->get_logger(), "Successed to enable FRAMOS");
    } else {
      RCLCPP_WARN(this->get_logger(), "Failed to enable FRAMOS");
    }
    is_reset_running_ = false;
  };

  auto disable_done_callback = [this, enable_done_callback](rclcpp::Client<rcl_interfaces::srv::SetParametersAtomically>::SharedFuture response) {
    if (response.valid() && response.get()->result.successful) {
      RCLCPP_INFO(this->get_logger(), "Successed to disable FRAMOS");

      std::vector<rcl_interfaces::msg::Parameter> params;
      rcl_interfaces::msg::Parameter rgb_param;
      rgb_param.name = "enable_color";
      rgb_param.value.type = rcl_interfaces::msg::ParameterType::PARAMETER_BOOL;
      rgb_param.value.bool_value = true;
      params.push_back(rgb_param);
      rcl_interfaces::msg::Parameter depth_param;
      depth_param.name = "enable_depth";
      depth_param.value.type = rcl_interfaces::msg::ParameterType::PARAMETER_BOOL;
      depth_param.value.bool_value = true;
      params.push_back(depth_param);

      auto request = std::make_shared<rcl_interfaces::srv::SetParametersAtomically::Request>();
      request->parameters = params;
      set_parameters_client_->async_send_request(request, enable_done_callback);
    } else {
      RCLCPP_WARN(this->get_logger(), "Failed to disable FRAMOS");
      is_reset_running_ = false;
    }
  };

  is_reset_running_ = true;

  time_tf_ready_ = 0.0;
  time_camera_ready_ = 0.0;
  std::queue<double>().swap(rgb_times_);
  std::queue<double>().swap(depth_times_);

  std::vector<rcl_interfaces::msg::Parameter> params;
  rcl_interfaces::msg::Parameter rgb_param;
  rgb_param.name = "enable_color";
  rgb_param.value.type = rcl_interfaces::msg::ParameterType::PARAMETER_BOOL;
  rgb_param.value.bool_value = false;
  params.push_back(rgb_param);
  rcl_interfaces::msg::Parameter depth_param;
  depth_param.name = "enable_depth";
  depth_param.value.type = rcl_interfaces::msg::ParameterType::PARAMETER_BOOL;
  depth_param.value.bool_value = false;
  params.push_back(depth_param);

  auto request = std::make_shared<rcl_interfaces::srv::SetParametersAtomically::Request>();
  request->parameters = params;
  set_parameters_client_->async_send_request(request, disable_done_callback);
}

void FramosInitializeNode::check_transform()
{
  if (is_ready_) {
    RCLCPP_INFO(this->get_logger(), "Framos is ready");
    rclcpp::shutdown();
    std::exit(EXIT_SUCCESS);
  }

  // check tf because launching FRAMOS before initial localization often fails on kx models
  double now = rclcpp::Time(get_clock()->now()).seconds();
  try {
    geometry_msgs::msg::TransformStamped tf_stamped = tf_buffer_->lookupTransform(
      map_frame_name_, camera_link_frame_name_,
      rclcpp::Time(0), rclcpp::Duration(std::chrono::duration<double>(1.0)));
    if (time_tf_ready_ <= 0) {
      time_tf_ready_ = now;
    }
  } catch (tf2::TransformException & ex) {
    RCLCPP_INFO(this->get_logger(), "TF is not ready");
    time_tf_ready_ = 0.0;
    return;
  }

  // calculate FPS
  int rgb_buffer_size = 0;
  int depth_buffer_size = 0;
  double rgb_fps = 0.0;
  double depth_fps = 0.0;
  {
    std::lock_guard<std::mutex> rgb_times_lock(rgb_times_mutex_);
    std::lock_guard<std::mutex> depth_times_lock(depth_times_mutex_);
    rgb_buffer_size = rgb_times_.size();
    depth_buffer_size = depth_times_.size();
    if ((rgb_buffer_size == max_buffer_size_) && (depth_buffer_size == max_buffer_size_)) {
      double rgb_buffer_time = rgb_times_.back() - rgb_times_.front();
      if (rgb_buffer_time > 0) {
        rgb_fps = rgb_buffer_size / rgb_buffer_time;
      }
      double depth_buffer_time = depth_times_.back() - depth_times_.front();
      if (depth_buffer_time > 0) {
        depth_fps = depth_buffer_size / depth_buffer_time;
      }
    }
  }

  if ((rgb_buffer_size == max_buffer_size_) && (depth_buffer_size == max_buffer_size_)) {
    // reset FRAMOS if FPS is lower than threshold
    if ((rgb_fps >= min_rgb_fps_) && (depth_fps >= min_depth_fps_)) {
      RCLCPP_INFO(this->get_logger(), "FRAMOS is ready, RGB FPS: %.2f, depth FPS: %.2f", rgb_fps, depth_fps);
      is_ready_ = true;
    } else {
      RCLCPP_WARN(this->get_logger(), "Reset FRAMOS because FPS is low, RGB FPS: %.2f, depth FPS: %.2f", rgb_fps, depth_fps);
      reset_framos();
    }
  } else if ((now - time_tf_ready_) > max_wait_after_tf_ready_) {
    // reset FRAMOS if buffers are not filled for a long time
    RCLCPP_WARN(this->get_logger(), "Reset FRAMOS because FRAMOS seems not working");
    reset_framos();
  }
}

}  // namespace cabot_people

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<cabot_people::FramosInitializeNode>(rclcpp::NodeOptions());
  while (rclcpp::ok()) {
    if (node->is_ready()) {
      rclcpp::shutdown();
      break;
    }
    rclcpp::spin_some(node);
  }
  return 0;
}
